topology2: Add rt5682+max98357a support

yongzhi1 opened the pull request:
make topology2_sof-mtl-max98357a-rt5682

localhost ~ # aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: sofrt5682 [sof-rt5682], device 0: Headset (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofrt5682 [sof-rt5682], device 1: dummy-spk (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofrt5682 [sof-rt5682], device 5: HDMI1 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofrt5682 [sof-rt5682], device 6: HDMI2 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofrt5682 [sof-rt5682], device 7: HDMI3 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofrt5682 [sof-rt5682], device 8: HDMI4 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
localhost ~ # arecord -l
**** List of CAPTURE Hardware Devices ****
card 0: sofrt5682 [sof-rt5682], device 0: Headset (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0

Signed-off-by: Yong Zhi <yong.zhi@intel.com>

plbossart commented on tools/topology/topology2/cavs/cavs-rt5682.conf:
```diff
@@ -0,0 +1,240 @@
+	}
+
+	pcm."1" {
+		name	"dummy-spk"
```
That doesn't seem like a great name.

What's wrong with 'Speakers'?

yongzhi1 replied:
Ack, MX98357A goes by "dummy speaker" cause it does not requires I2C programming :)

plbossart commented on tools/topology/topology2/cavs/CMakeLists.txt:
```diff
@@ -44,6 +44,11 @@ PREPROCESS_PLUGINS=nhlt,NHLT_BIN=nhlt-cavs-tgl-nocodec.bin"
+
+# SSP topology for MTL
+"cavs-rt5682\;sof-mtl-max98357a-rt5682\;PLATFORM=mtl,\
+NUM_HDMIS=4,PREPROCESS_PLUGINS=nhlt,NHLT_BIN=nhlt-ace-mtl-nocodec.bin"
```
Zero dmics on this platform?

yongzhi1 replied:
There are two reasons I did not put DMIC
1. The build will fail as reported in https://github.com/thesofproject/sof/issues/6007
2. On our test RVP with I2S config, DMIC option is not available, so it does not hurt.

Maybe I can change sof-mtl-max98357a-rt5682 to sof-mtl-max98357a-rt5682-no-dmic?

plbossart replied:
@yongzhi1 
a) we should fix known errors first, not work-around them
b) if this is a test topology, it needs to go in a 'development' folder, so that it's clear we don't intend to release it to the general public. A device without microphone isn't very useful, is it? Well, unless it's a feature for security/privacy usages....

yongzhi1 replied:
Got it, let me wait for one week and hopefully we can figure out the build issue, if not, will send it to development folder.

marc-hb replied:
> figure out the build issue, if not, will send it to development folder.

Do you mean temporarily, as staging area until it's fixed? Most other topologies in the development folder will _never_ be released in https://github.com/thesofproject/sof-bin because they're not meant to be used in production on any device.

The difference between the development folder and the other, released folder is usually not a "quality" difference. It's usually a difference between something used in production on someone's device versus not.

yongzhi1 replied:
> > figure out the build issue, if not, will send it to development folder.
> 
> Do you mean temporarily, as staging area until it's fixed? Most other topologies in the development folder will _never_ be released in https://github.com/thesofproject/sof-bin because they're not meant to be used in production on any device.
> 
> The difference between the development folder and the other, released folders is usually not a "quality" difference. It's usually a difference between something used in production on customer's device versus not.

Then the development folder should be the right place since the tplg is intended for RVP only not product, think the nocodec tplg etc. should also be moved to the same folder.

bardliao commented on tools/topology/topology2/cavs/cavs-rt5682.conf:
```diff
@@ -0,0 +1,240 @@
+# Pipeline ID:1 PCM ID: 0
+Object.Pipeline {
+	# playback pipelines
+	passthrough-playback."1" {
```
We don't need the quotation marks. ie. Please use `passthrough-playback.1 {` like others

yongzhi1 replied:
Ack, will update in next version.

lgirdwood commented:
@yongzhi1 any update ?

yongzhi1 commented:
V2:
-  updated PCM.1 name from "dummy-spk" to "Speakers"
-  updated NHLT_BIN name to nhlt-ace-mtl-rt5682.bin
-  removed quote around pipeline_id  "N" 
-  added DMIC pipelines
-  *change SSP1's direction to duplex instead of playback

For the last change, not sure if this is a bug, without the change, capture on SSP0 will fail for:

[  255.224147] sof-audio-pci-intel-mtl 0000:00:1f.3: no matching blob for sample rate: 48000 sample width: 32 channels: 2
[  255.236138] sof-audio-pci-intel-mtl 0000:00:1f.3: failed to prepare widget copier.SSP.8.1
[  255.245289] sof-audio-pci-intel-mtl 0000:00:1f.3: Failed to prepare connected widgets
[  255.254045] sof-audio-pci-intel-mtl 0000:00:1f.3: error: failed widget list set up for pcm 0 dir 1
[  255.264093] sof-audio-pci-intel-mtl 0000:00:1f.3: ASoC: error at snd_soc_pcm_component_hw_params on 0000:00:1f.3: -22
[  255.276045]  Headset: ASoC: __soc_pcm_hw_params() failed (-22)
[  255.282591]  Headset: ASoC: dpcm_fe_dai_hw_params failed (-22)

yongzhi1 commented:
V3
- changed SSP1 direction back from "duplex" to "playback"
(tplg build and tested with https://github.com/juimonen/alsa-utils/tree/dai_index_fix)

lgirdwood commented:
@yongzhi1 does this pend on any recent unmerged fixes from @juimonen ?

yongzhi1 commented:
> @yongzhi1 does this pend on any recent unmerged fixes from @juimonen ?

Yes, that's correct - it depends on the dai_index fix to generate the working tplg, but the conf itself is OK :)

plbossart commented on tools/topology/topology2/cavs/CMakeLists.txt:
```diff
@@ -44,6 +44,12 @@ PREPROCESS_PLUGINS=nhlt,NHLT_BIN=nhlt-cavs-tgl-nocodec.bin"
+# SSP topology for MTL
+"cavs-rt5682\;sof-mtl-max98357a-rt5682\;PLATFORM=mtl,\
+NUM_DMICS=4,PDM1_MIC_A_ENABLE=1,PDM1_MIC_B_ENABLE=1,\
+NUM_HDMIS=4,PREPROCESS_PLUGINS=nhlt,NHLT_BIN=nhlt-ace-mtl-rt5682.bin"
```
I am still very very confused with the NHLT naming. We need separate SSP blobs for rt5682 and max98357a, and there's no mention on what the dmic configuration is. That's a recipe for confusion and possible disaster when reusing stuff.

yongzhi1 replied:
Do you mean NHLT_BIN? true, any established NHLT naming convention that I can follow?

bardliao replied:
@plbossart IIRC, the NHLT_BIN is the hnlt blob of this topology. The purpose is that we can analyze the nhlt bin file if we want to debug the nhlt blob. So the nhlt bin is associated to the .tplg file. Maybe we can always use the same name as the .tplg file? Like sof-mtl-max98357a-rt5682.tplg and nhlt-sof-mtl-max98357a-rt5682.bin.
@ranj063 @juimonen Please correct me if I was wrong.

juimonen replied:
Yes the .bin is debug feature to extract nhlt blob from the process (so you don't have to dig it out from the topology binary). Not sure should we have some general DEBUG_TOPOLOGY flag in CMake. I don't personally mind it being there, but others might when the amount of topologies grows.

lgirdwood replied:
This NHLT needs to be referenced by source file name or by coding the config in the topology (when we are not reading from ACPI).

plbossart commented on tools/topology/topology2/cavs/cavs-rt5682.conf:
```diff
@@ -0,0 +1,240 @@
+Define {
+	MCLK 				24576000
+	NUM_DMICS			0
+	SPK_ID				7
```
what is that?

yongzhi1 replied:
SPK_ID is the HDMI4_ID + 1, NUM_DMICS and MCLK are the default values can be overwritten.

plbossart commented on tools/topology/topology2/cavs/cavs-rt5682.conf:
```diff
@@ -0,0 +1,240 @@
+
+# include HDMI config if needed.
+IncludeByKey.NUM_HDMIS {
+"[3-4]" "platform/intel/hdmi-generic.conf"
```
indentation

yongzhi1 replied:
Ack.

plbossart commented on tools/topology/topology2/cavs/cavs-rt5682.conf:
```diff
@@ -0,0 +1,240 @@
+
+	passthrough-playback.2 {
+		index	3
+		Object.Widget.pipeline.1.stream_name	"passthrough-playback-ssp1"
```
For all MTL-based products, we want the ability to support multiple streams for playback and capture. So we cannot have any references to passthrough any longer. We need a clear split between host and dai pipelines, so that additional host pipelines can be added at a later point.

yongzhi1 replied:
Ok, then let me digest above, thanks.

gkbldcig commented:
Can one of the admins verify this patch?

yongzhi1 commented:
V4 update:

- Rename NHL bin file.
- Use volume/mixer for SSP playback.

Just try to bring the code closer to its final form even thought the maturity of mixer is not ready.

lgirdwood commented:
@plbossart @juimonen @bardliao any comments on the updates ?

bardliao reviewed (approved):
LGTM

yongzhi1 commented:
V5:

Cosmetic updates.

plbossart commented on tools/topology/topology2/cavs/cavs-rt5682.conf:
```diff
@@ -0,0 +1,274 @@
+# PCM1 ---> gain ----> Mixin ----> Mixout ----> gain ----> SSP1
+#
+# SSP0 ----> PCM0
+# SSP1 ----> PCM1
```
this inconsistent with the multiple-stream requirement on capture. Is the plan to add additional copiers and PCMs later for capture?

yongzhi1 replied:
Yes, plan to add DeepBuffer and multi-stream later later incrementally, also re-use the same file to support Rex0 variant, thank you for review.

lgirdwood commented:
@yongzhi1 can you confirm the version of alsatplg you are using. It looks like the CI could be using an older version ?
```
[2/709] Generating sof-mtl-max98357a-rt5682.tplg
find_modes(): fs not set
alsatplg: malloc.c:2379: sysmalloc: Assertion `(old_top == initial_top (av) && old_size == 0) || ((unsigned long) (old_size) >= MINSIZE && prev_inuse (old_top) && ((unsigned long) old_end & (pagesize - 1)) == 0)' failed.
Aborted (core dumped)
```

yongzhi1 commented:
> @yongzhi1 can you confirm the version of alsatplg you are using. It looks like the CI could be using an older version ?
> 
> ```
> [2/709] Generating sof-mtl-max98357a-rt5682.tplg
> find_modes(): fs not set
> alsatplg: malloc.c:2379: sysmalloc: Assertion `(old_top == initial_top (av) && old_size == 0) || ((unsigned long) (old_size) >= MINSIZE && prev_inuse (old_top) && ((unsigned long) old_end & (pagesize - 1)) == 0)' failed.
> Aborted (core dumped)
> ```

Hi, @lgirdwood , thank you for noticing, how to get https://github.com/alsa-project/alsa-utils/pull/162 merged?

ranj063 commented:
> @ranj063 @juimonen should we send this patch to the mailing list instead ?

@lgirdwood alsaproject PR's always generate an email to the mailing list as well. In the past we've always pinged @perexg on the PR's to get his attention and looks like Yong has already done it.

The more important thing though would be to update our CI docker with the change in alsautils

yongzhi1 commented:
Thank you @ranj063 for the hint, the alsa-utils patch got merged :)

lgirdwood commented:
> Thank you @ranj063 for the hint, the alsa-utils patch got merged :)

@marc-hb can you update the container with latest alsa-utils git when convenient. Thanks !

yongzhi1 commented:
v6:

Added default value for USE_CHAIN_DMA to fix build err.

yongzhi1 commented:
v7:

Rebase due to topology dir structure changes.
